File: node_depth/nodeDepth.py
#Complexity: O(n) n - number of nodes
#Space: O(h) h - height of tree (for stack)
def nodeDepths(root):
    res = 0
    stack = [{"node":root, "depth":0}]
    while len(stack):
        tmp = stack.pop()
        node, depth = tmp["node"], tmp["depth"]
        if not node:
            continue
        res += depth
        stack.append({"node": node.left, "depth": depth + 1})
        stack.append({"node": node.right, "depth": depth + 1})
    return res


class BinaryTree:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

    def insert(self, value):
        if value < self.value:
            if not self.left:
                self.left = BinaryTree(value)
            else:
                self.left.insert(value)
        else:
            if not self.right:
                self.right = BinaryTree(value)
            else:
                self.right.insert(value)

    def __str__(self, level=0):
        res = "\t"*level+str(self.value)+"\n"
        if self.left:
            res += self.left.__str__(level+1)
        if self.right:
            res += self.right.__str__(level+1)
        return res


if __name__ == '__main__':
    tree = BinaryTree(5)
    tree.insert(4)
    tree.insert(3)
    tree.insert(2)
    tree.insert(1)
    tree.insert(6)
    tree.insert(7)
    tree.insert(8)
    tree.insert(9)
    tree.insert(10)
    print(tree)
    print("25 vs", nodeDepths(tree))

File: node_depth/nodeDepth.go
package main

import (
	"fmt"
	"io"
	"os"
)

type Node struct {
	Value int
	Left  *Node
	Right *Node
}

type BST struct {
	Root *Node
}

func calcDepthSumHelper(node *Node, sums *[]int, lvl int) {
	if node == nil {
		return
	}
	*sums = append(*sums, lvl)
	calcDepthSumHelper(node.Left, sums, lvl+1)
	calcDepthSumHelper(node.Right, sums, lvl+1)
}

func calcDepthSum(tree *BST) int {
	var res int
	var sums []int

	calcDepthSumHelper(tree.Root, &sums, 0)
	for _, num := range sums {
		res += num
	}
	return res
}

func (node *Node) insert(value int) {
	if node == nil {
		return
	} else if value < node.Value {
		if node.Left == nil {
			node.Left = &Node{Value: value, Left: nil, Right: nil}
		} else {
			node.Left.insert(value)
		}
	} else {
		if node.Right == nil {
			node.Right = &Node{Value: value, Left: nil, Right: nil}
		} else {
			node.Right.insert(value)
		}
	}
}

func (tree *BST) insert(value int) *BST {
	if tree.Root == nil {
		tree.Root = &Node{Value: value, Left: nil, Right: nil}
	} else {
		tree.Root.insert(value)
	}
	return tree
}

func print(w io.Writer, node *Node, ns int, ch rune) {
	if node == nil {
		return
	}
	for i := 0; i < ns; i++ {
		fmt.Fprint(w, " ")
	}
	fmt.Fprintf(w, "%c:%v\n", ch, node.Value)
	print(w, node.Left, ns+2, 'L')
	print(w, node.Right, ns+2, 'R')
}

func main() {
	tree := &BST{}
	tree.insert(5).
		insert(4).
		insert(3).
		insert(2).
		insert(1).
		insert(6).
		insert(7).
		insert(8).
		insert(9).
		insert(10)
	print(os.Stdout, tree.Root, 0, 'M')
	summa := 25
	fmt.Println(summa, "vs", calcDepthSum(tree))
}
